generator/complex_generator/CMakeLists.txt

```diff
@@ -6,5 +6,5 @@ omim_add_executable(${PROJECT_NAME} ${SRC})
 
 target_link_libraries(${PROJECT_NAME}
   generator
-  gflags::gflags
```

Comment:
Maybe use `set(GFLAGS_USE_TARGET_NAMESPACE ON)` instead of changing all these files? https://github.com/gflags/gflags/blob/a738fdf9338412f83ab3f26f31ac11ed3f3ec4bd/cmake/config.cmake.in#L13

Reply:
Hmm, it seems not to work. I have applied the following commit instead of changing all files:
```diff

diff --git a/3party/CMakeLists.txt b/3party/CMakeLists.txt                                  
index a54a1ecc4b..a30bb20c25 100644                                                         
--- a/3party/CMakeLists.txt                                                                 
+++ b/3party/CMakeLists.txt                                                                 
@@ -2,7 +2,10 @@                                                                            
 # Compatibility with CMake < 3.5 will be removed from a future version of CMake.
 set(CMAKE_WARN_DEPRECATED OFF CACHE BOOL "" FORCE)

-if (NOT WITH_SYSTEM_PROVIDED_3PARTY)                                                       
+if (WITH_SYSTEM_PROVIDED_3PARTY)                                                           
+  set(GFLAGS_USE_TARGET_NAMESPACE ON)                                                      
+  find_package(gflags REQUIRED)                                                            
+else()                                                                                     
   # Configure expat library.
   # Suppress "Policy CMP0077 is not set: option() honors normal variables"
   # for the expat options below.
```
I also tried the lines (setting the variable and find the package) the other way around.

Both times, this was the error message (the same as if I would not set `GFLAGS_USE_TARGET_NAMESPACE`):
```
CMake Error at map/benchmark_tool/CMakeLists.txt:12 (target_link_libraries):
  Target "benchmark_tool" links to:

    gflags::gflags

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.
```
My system cmake files of gflags (version 2.2.2) include the same lines than the one you linked.

Reply:
Strange, I also made a small test project to reproduce the issue. In this test project your workaround works:
```cmake
cmake_minimum_required(VERSION 3.18)
project(gflags-test CXX)

# get_property(importTargets DIRECTORY "${CMAKE_SOURCE_DIR}" PROPERTY IMPORTED_TARGETS)

set(GFLAGS_USE_TARGET_NAMESPACE ON)
find_package(gflags REQUIRED)

# get_property(importTargetsAfter DIRECTORY "${CMAKE_SOURCE_DIR}" PROPERTY IMPORTED_TARGETS)
# list(REMOVE_ITEM importTargetsAfter ${importTargets})
# message("${importTargetsAfter}")

add_executable(gflags-test gflags-test.cpp)

target_link_libraries(gflags-test
  gflags::gflags
)
```

I will try it another time with Organic Maps.

Reply:
More debugging on that:

## Minimal example (from the post above)

When uncommenting the `get_property` lines, this is the CMake output:
```
[0/1] Re-running CMake...
gflags::gflags_shared;gflags::gflags_nothreads_shared;gflags::gflags
-- Configuring done (0.0s)
-- Generating done (0.0s)
-- Build files have been written to: /home/gerion/test/gflags-user/build
ninja: no work to do.
```
So, everything works fine, the `gflags::gflags` target is found.

## Organic Maps

I then added this to Organic Maps (in addition to the patch above):
```diff
diff --git a/3party/CMakeLists.txt b/3party/CMakeLists.txt
index 70e1614abf..94545c3abd 100644
--- a/3party/CMakeLists.txt
+++ b/3party/CMakeLists.txt
@@ -3,8 +3,19 @@
 set(CMAKE_WARN_DEPRECATED OFF CACHE BOOL "" FORCE)
 
 if (WITH_SYSTEM_PROVIDED_3PARTY)
+  message("imported begin")
+  get_property(importTargets DIRECTORY "${CMAKE_SOURCE_DIR}" PROPERTY IMPORTED_TARGETS)
+
   set(GFLAGS_USE_TARGET_NAMESPACE ON)
   find_package(gflags REQUIRED)
+
+  get_property(importTargetsAfter DIRECTORY "${CMAKE_SOURCE_DIR}" PROPERTY IMPORTED_TARGETS)
+  list(REMOVE_ITEM importTargetsAfter ${importTargets})
+  message("${importTargetsAfter}")
+  message("imported end")
+  if (NOT TARGET gflags::gflags)
+    message (FATAL_ERROR "gflags not found")
+  endif ()
 else()
   # Configure expat library.
   # Suppress "Policy CMP0077 is not set: option() honors normal variables"
diff --git a/map/benchmark_tool/CMakeLists.txt b/map/benchmark_tool/CMakeLists.txt
index e8dc83b93a..206f3ee707 100644
--- a/map/benchmark_tool/CMakeLists.txt
+++ b/map/benchmark_tool/CMakeLists.txt
@@ -9,6 +9,10 @@ set(SRC
 
 omim_add_executable(${PROJECT_NAME} ${SRC})
 
+if (NOT TARGET gflags::gflags)
+  message (FATAL_ERROR "gflags in map not found")
+endif ()
+
 target_link_libraries(${PROJECT_NAME}
   map
   gflags::gflags
```
So, after importing gflags, CMake should print out the newly found imported targets, fail if the target is not found, and again fail when the target is not found in the map/benchmark_tool subproject.

This is the CMake output:
```
% cmake .. -GNinja -DWITH_SYSTEM_PROVIDED_3PARTY=yes
-- Using compiler GNU 13.2.1
-- Using Unity Build with batch 50, export UNITY_DISABLE=1 or use -DUNITY_DISABLE=ON to disable it.
-- Using ccache, export CCACHE_DISABLE=1 or use -DCCACHE_DISABLE=ON to disable it.
-- export COLORS_DISABLE=1 or use -DCOLORS_DISABLE=ON to disable colored compiler output.
Setting PLATFORM_LINUX to true
-- Build type: Release
-- Using ld.ldd linker
-- Found the following ICU libraries:
--   uc (required): /usr/lib64/libicuuc.so
--   i18n (required): /usr/lib64/libicui18n.so
--   data (required): /usr/lib64/libicudata.so
imported begin

imported end
-- Found python to use in qt/, shaders/ and 3party/: /usr/bin/python3.12
CMake Error at map/benchmark_tool/CMakeLists.txt:13 (message):
  gflags in map not found


-- Configuring incomplete, errors occurred!
```
So, `gflags::gflags` is _not_ an imported target but a target but only when not used in a subproject. I'm not that skilled in CMake. Do you have a clue what the problem is?

Additional hint: I found out with printf debugging, that the following lines from the upstream gflags cmake config are executed:
```cmake
# add more convenient "${GFLAGS_TARGET_PREFIX}gflags" import target
if (NOT TARGET ${GFLAGS_TARGET_PREFIX}gflags)
  if (GFLAGS_SHARED)
    add_library (${GFLAGS_TARGET_PREFIX}gflags SHARED IMPORTED)
  else ()
    add_library (${GFLAGS_TARGET_PREFIX}gflags STATIC IMPORTED)
  endif ()
  ...
```
`GFLAGS_SHARED` is true and `GFLAGS_TARGET_PREFIX` is `gflags::`.

Reply:
Try clearing cmake cache. Also just in case check whether there are other calls to `find_package(gflags)`. Also try calling `find_package(gflags REQUIRED)` at the root `CMakeLists.txt`.

Reply:
> Try clearing cmake cache.

Does not help.

> Also just in case check whether there are other calls to `find_package(gflags)`.

They do not exist.

> Also try calling `find_package(gflags REQUIRED)` at the root `CMakeLists.txt`.

That helped. I changed my patch accordingly.